js/compiler.ts

```diff
@@ -263,53 +389,80 @@ export class DenoCompiler implements ts.LanguageServiceHost {
     );
 
     const [outputFile] = output.outputFiles;
-    return outputFile.text;
+    const outputCode = (moduleMetaData.outputCode = `${
+      outputFile.text
+    }\n//# sourceURL=${fileName}`);
+    moduleMetaData.scriptVersion = "1";
+    this._os.codeCache(fileName, sourceCode, outputCode);
+    return moduleMetaData.outputCode;
+  }
+
+  /**
+   * For a given module specifier and containing file, return a list of absolute
+   * identifiers for dependent modules that are required by this module.
+   */
+  getModuleDependencies(
+    moduleSpecifier: ModuleSpecifier,
+    containingFile: ContainingFile
+  ): ModuleFileName[] {
+    assert(
+      this._instantiateQueue.length === 0,
+      "Cannot get dependencies with uninstantiated modules."
+    );
+    const moduleMetaData = this.resolveModule(moduleSpecifier, containingFile);
+    assert(
+      !moduleMetaData.hasRun,
+      "Cannot get dependencies for an already instantiated module."
+    );
+    this._resolveDependencies(moduleMetaData);
+    const dependencies = this._instantiateQueue.map(
+      moduleMetaData => moduleMetaData.fileName
+    );
+    this._instantiateQueue = [];
+    return dependencies;
   }
 
   /**
    * Create a localized AMD `define` function and return it.
    */
   makeDefine(moduleMetaData: ModuleMetaData): AmdDefine {
-    const localDefine = (deps: string[], factory: AmdFactory): void => {
-      // TypeScript will emit a local require dependency when doing dynamic
-      // `import()`
-      const { _log: log } = this;
-      const localExports = moduleMetaData.exports;
-
-      // tslint:disable-next-line:no-any
-      const resolveDependencies = (deps: string[]): any[] => {
-        return deps.map(dep => {
-          if (dep === "require") {
-            return localRequire;
-          } else if (dep === "exports") {
-            return localExports;
-          } else if (dep in DenoCompiler._builtins) {
-            return DenoCompiler._builtins[dep];
-          } else {
-            const depModuleMetaData = this.run(dep, moduleMetaData.fileName);
-            return depModuleMetaData.exports;
-          }
-        });
-      };
-
-      // this is a function because we need hoisting
-      function localRequire(
-        deps: string[],
-        callback: AmdCallback,
-        errback: AmdErrback
-      ): void {
-        log("localRequire", deps);
-        try {
-          const args = resolveDependencies(deps);
-          callback(...args);
-        } catch (e) {
-          errback(e);
+    // TODO should this really be part of the public API of the compiler?
```

Comment:
Can you elaborate on this TODO, I'm not sure I understand?

Reply:
They were not private in previous versions, so since I wasn't specifically needing to change the public API of the compiler, I put that there. Upon reflection, I think `resolveModule` `run` `compile` and `getModuleDependencies` are only valid compiler APIs that should be part of the public API surface (except of course the TypeScript ones we need), but also it needs to bake longer so I just put a note there for now.